message/pool/pool.go

```diff
@@ -0,0 +1,55 @@
+package pool
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/atomic"
+)
+
+type Pool struct {
+	// This field needs to be the first in the struct to ensure proper word alignment on 32-bit platforms.
+	// See: https://golang.org/pkg/sync/atomic/#pkg-note-BUG
+	currentMessagesInPool atomic.Int64
+	messagePool           sync.Pool
+	maxNumMessages        uint32
+	maxMessageBufferSize  uint16
+}
+
+func New(maxNumMessages uint32, maxMessageBufferSize uint16) *Pool {
+	return &Pool{
+		maxNumMessages:       maxNumMessages,
+		maxMessageBufferSize: maxMessageBufferSize,
+	}
+}
+
+// AcquireMessage returns an empty Message instance from Message pool.
+//
+// The returned Message instance may be passed to ReleaseMessage when it is
+// no longer needed. This allows Message recycling, reduces GC pressure
+// and usually improves performance.
+func (p *Pool) AcquireMessage(ctx context.Context) *Message {
+	v := p.messagePool.Get()
+	if v == nil {
+		return NewMessage(ctx)
+	}
+	r := v.(*Message)
+	p.currentMessagesInPool.Dec()
```

Comment:
Wouldn't this become negative when you invoke `AcquireMessage` twice before invoking `ReleaseMessage`? In that case the counter wouldn't work.

Example:
```
p := pool.New(1, 1024)

p.AcquireMessage(ctx) // p.currentMessagesInPool == -1
p.AcquireMessage(ctx) // p.currentMessagesInPool == -2
p.ReleaseMessage(req) // p.currentMessagesInPool == -1, req is put into the pool
p.ReleaseMessage(req) // p.currentMessagesInPool == 0, req is put into the pool, so more messages are in the pool than allowed
```

Reply:
No. `p.currentMessagesInPool` is incremented when message is released.
```
p := pool.New(1, 1024)
p.AcquireMessage(ctx) // p.currentMessagesInPool = 0
p.AcquireMessage(ctx) // p.currentMessagesInPool = 0
p.AcquireMessage(ctx) // p.currentMessagesInPool = 0
p.ReleaseMessage(req) // p.currentMessagesInPool = 1
```

Reply:
`p.messagePool.Get()` will return nil, because no message has been put into the pool yet, right? Indeed, this should be correct then.